R/xgboost-deployment.R

```diff
@@ -0,0 +1,281 @@
+      # R factors are 1 indexed, XGB is 0 indexed, so we must subtract 1 from the labels. They must be numeric.
+      temp_test_data <- data.matrix(private$dfTestTemp[ ,!(colnames(private$dfTestTemp) == self$params$predictedCol)])
+      temp_test_label <- data.matrix(as.numeric(private$dfTestTemp[[self$params$predictedCol]])) - 1 
+      self$xgb_testMatrix <- xgb.DMatrix(data = temp_test_data, label = temp_test_label) 
+      private$test_label <- temp_test_label # save for confusion matrix and output
+      rm(temp_test_data, temp_test_label) # clean temp variables
+    },
+
+    # Perform prediction
+    performPrediction = function() {
+      cat('Generating Predictions...','\n')
+      private$temp_predictions <- predict(object = private$fitXGB,
+                                  newdata = self$xgb_testMatrix,
+                                  reshape = TRUE)
+      
+      # Build prediction output
+      private$predictions <- private$temp_predictions %>% 
+        data.frame() %>%
+        mutate(predicted_label = max.col(.),
+               true_label = private$test_label + 1)
+
+      # Set column names to match input targets
+      colnames(private$predictions)[1:self$params$xgb_numberOfClasses] <- self$params$xgb_targetNames
+      colnames(private$temp_predictions)[1:self$params$xgb_numberOfClasses] <- self$params$xgb_targetNames
+
+      # Set up a mapping for the values themselves
+      from <- 1:self$params$xgb_numberOfClasses
+      to <- self$params$xgb_targetNames
+      map = setNames(to,from)
+      private$predictions$predicted_label <- map[private$predictions$predicted_label] # note square brackets
+      private$predictions$true_label <- map[private$predictions$true_label] 
+
+      # Prepare output 
+      private$predictions <- cbind(private$grainTest, private$predictions)
+      colnames(private$predictions)[1] <- self$params$grainCol
+      
+      if (isTRUE(self$params$debug)) {
+        cat('Rows in multiclass prediction: ', length(private$predictions), '\n')
+        cat('First 10 raw probability predictions (with row # first)', '\n')
+        print(private$predictions[1:10,])
+      }
+    },
+
+    calculateOrderedFactors = function() {
+      cat('Ordering top probabilities...', '\n')
+      # Calculate ordered factors of importance for each row's prediction
+      # Get column indices of max values in each row
+      nRows = dim(private$temp_predictions)[1]
+      maxColInds <- t(sapply(1:nRows, function(i) 
+        order(private$temp_predictions[i,], decreasing=TRUE)))
+      
+      # sort column values in each row
+      maxColVals <- t(sapply(1:nRows, function(i) 
+        private$temp_predictions[i,maxColInds[i,]]))
+
+      # get names of maximum values for each row
+      maxColNames <- t(sapply(1:nRows, 
+        function(i) colnames(private$temp_predictions)[maxColInds[i,]]))
+
+      # combine top 3 maxColVals and maxColNames to make maxProbDF
```

Comment:
-Put a note in example that says output of boostD$getPredictions() gives 3 highest probabilities and their associated classes.
OR
-Perhaps give "PredictedProb1" and in the next column give/say "PredictedProb1Class"?